macrosynergy/learning/predictors.py

```diff
@@ -1,16 +1,391 @@
+import inspect
+
 import numpy as np
+import pandas as pd
+
+import datetime
+
 from sklearn.base import BaseEstimator, RegressorMixin
 
+from typing import Union
+
 
 class NaivePredictor(BaseEstimator, RegressorMixin):
     """
     Naive predictor class to output a column of features as the predictions themselves.
     It expects only one column in the input matrix for which predictions are to be made.
     This class would typically be used in the final stage of a `scikit-learn` pipeline,
-    following either `FeatureAverager` or `ZnScoreAverager` stages. 
+    following either `FeatureAverager` or `ZnScoreAverager` stages.
     """
+
     def fit(self, X, y=None):
         return self
 
     def predict(self, X):
         return np.array(X)
+
+
+class SignWeightedRegressor(BaseEstimator, RegressorMixin):
+    def __init__(self, regressor: RegressorMixin, **init_kwargs):
+        """
+        Custom predictor class to create a regression model where different training samples
+        are weighted differently depending on the sign of their associated targets. If there
+        are more positive targets than negative targets in the training set, then the
+        negative target samples are given a higher weight in the model training process.
+        The opposite is true if there are more negative targets than positive targets.
+
+        :param <RegressorMixin> regressor: Scikit-learn regressor class to be used as the
+            underlying model for which a sign-weighted fit is to be performed.
+
+        NOTE: This class is specifically tailored for compatibility with any scikit-learn
+        regressor that supports a 'sample_weight' parameter in its 'fit' method. Each
+        training sample receives a weight inversely proportional to the frequency of its
+        label's sign in the training set. Thus, the model is encouraged to learn equally
+        from both positive and negative return samples.
+        """
+        if not callable(regressor) or not inspect.isclass(regressor):
```

Comment:
Can this and the next check be replaced by `if not isinstance(regressor, RegressorMixin) or not isinstance(regressor, BaseEstimator)`?

Reply:
No. Otherwise, LinearRegression and LinearRegression() would both pass but the second is not valid here -> for this class, we want all hyperparameters for LinearRegression (or another estimator) to be passed separately to allow for hyperparameter optimisation of those whilst using a weighted loss function.

Reply:
Ah I see!